elements/pl-external-grader-variables/pl-external-grader-variables.py

```diff
@@ -1,34 +1,90 @@
-    if params_name is None:
-        raise Exception('Attribute "params-name" is not defined.')
-    if params_name not in data["params"]:
-        raise Exception(
-            f"Variable name {params_name} does not exist in data['params']."
-        )
+    params_name = pl.get_string_attrib(element, "params-name")
 
+    # Get frontend variables
+    html_variables = []
 
-def render(element_html, element_index, data):
+    for child in element:
+        if child.tag == "pl-variable":
+            pl.check_attribs(child, ["name", "type"], [])
+
+            var_dict = {
+                "name": pl.get_string_attrib(child, "name"),
+                "type": pl.get_string_attrib(child, "type"),
+            }
+
+            description_html = pl.inner_html(child)
+
+            # Check for empty string, inner_html doesn't return None
+            if description_html:
+                var_dict["description"] = description_html
+
+            html_variables.append(var_dict)
+
+        elif child.tag is lxml.etree.Comment:
+            continue
+
+        else:
+            raise ValueError(
+                f"Tags inside of pl-external-grader-variables must be pl-variable, not '{child.tag}'."
+            )
+
+    declared_empty = pl.get_boolean_attrib(element, "empty", EMPTY_DEFAULT)
+
+    if declared_empty:
+        if html_variables:
+            raise ValueError(
+                f"Variable name '{params_name}' was declared empty, but has variables defined in 'question.html'."
+            )
+        elif params_name in data["params"]:
+            raise ValueError(
+                f"Variable name '{params_name}' was declared empty, but has variables defined in 'server.py'."
+            )
+
+        data["params"][params_name] = []
+    elif params_name not in data["params"]:
+        if not html_variables:
+            raise ValueError(
+                f"Variable name '{params_name}' has no variables defined in 'question.html' or 'server.py'. "
+                "Did you mean to set it to be empty?"
+            )
+
+        data["params"][params_name] = html_variables
+    else:
+        if html_variables:
+            raise ValueError(
+                f"Cannot define variables from both 'question.html' and 'server.py' for variable name '{params_name}'."
+            )
+
+
+def render(element_html: str, data: pl.QuestionData) -> str:
     element = lxml.html.fragment_fromstring(element_html)
-    params_name = pl.get_string_attrib(element, "params-name", PARAMS_NAME_DEFAULT)
+
+    params_name = pl.get_string_attrib(element, "params-name")
+
+    # Get final variables list
     names_user_description = data["params"][params_name]
     has_names_user_description = len(names_user_description) > 0
 
+    # Show descriptions if any variable has them set
+    has_descriptions = any(
+        "description" in var_description for var_description in names_user_description
```

Comment:
To go along with my other suggested change:
```suggestion
        d.get("description", None) for d in names_user_description
```